src/sfincs_bmi.cxx

```diff
@@ -11,239 +11,349 @@
 
 #include <cstring>
 #include <iostream>
+
 #include "sfincs_bmi.hxx"
 
 // sfincs_bmi.f90 implements/exports the following bmi functions:
-extern "C" int initialize(char *c_config_file);
-extern "C" int update(double dt);  // Doesn't seem to update current time
+extern "C" int initialize(const char *c_config_file);
 extern "C" int finalize();
-extern "C" void get_start_time(double *tstart);
-extern "C" void get_end_time(double *tend);
-extern "C" void get_time_step(double *dt);
-extern "C" void get_current_time(double *tcurrent);
-extern "C" void get_var_type(const char *c_var_name, char *c_type);
+
+extern "C" int get_component_name(char *name);
+
+extern "C" int update();
+extern "C" int update_until(double t);
+
+extern "C" int get_start_time(double *tstart);
+extern "C" int get_end_time(double *tend);
+extern "C" int get_time_step(double *dt);
+extern "C" int get_time_units(char *units);
+extern "C" int get_current_time(double *tcurrent);
+
+extern "C" int get_var_type(const char *c_var_name, char *c_type);
 
 // These function are also exported but not in the BMI spec
-extern "C" void get_var(const char *c_var_name, void *x);  // not exported; should be get_value?
-extern "C" void set_var(const char *c_var_name, float *xptr);  // should be set get_value?
-extern "C" void get_var_shape(const char *c_var_name, int *var_shape); // should be get_grid_shape
-// extern "C" void get_var_rank(char *c_var_name, int *rank); // should be get_grid_rank
+extern "C" int get_value(const char *c_var_name, void *x); // not exported; should be get_value?
+extern "C" int get_value_at_indices(const char *name, void *dest, int *inds, int count);
+extern "C" int get_value_ptr(const char *name, void **dest_ptr);
+
+extern "C" int set_value(const char *c_var_name, float *xptr); // should be set get_value?
+extern "C" int set_value_at_indices(const char *name, int *inds, int count, void *src);
+
+extern "C" int get_var_shape(const char *c_var_name, int *var_shape); // should be get_grid_shape
```

Comment:
BMI doesn't specify `get_var_shape`. Instead, it defines `get_grid_shape` for a `grid_id` that you can get through `get_var_grid`. I tried to implement a workaround in https://github.com/eWaterCycle/sfincs-bmi-server/pull/8/commits/add99b281fb4994e67f9aefe7704796899abf246

However, it seems that the `get_var_shape` implementation in fortran is not returning what we expect. According to [bmi spec](https://bmi.readthedocs.io/en/stable/#get-grid-shape) (also see [get_grid_rank](https://bmi.readthedocs.io/en/stable/#get-grid-rank)) this should return an array of len 2 (in case of a rank 2 array). In the fortran code, it seems a len(1) array is used for `var_shape` instead.

Note: we use `get_grid_shape` to transform the 1d array that we get from `get_value` back to a 2d array